shared/src/akri/instance.rs

```diff
@@ -323,6 +323,108 @@ fn default_shared() -> bool {
+    #[derive(Debug, PartialEq, Eq)]
+    pub struct ParseDeviceUsageError;
+    #[derive(PartialEq, Clone, Debug, Default)]
+    pub struct DeviceUsage {
+        kind: DeviceUsageKind,
+        usage_name: String,
+    }
+
+    impl std::fmt::Display for DeviceUsage {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            match &self.kind {
+                DeviceUsageKind::Free => write!(f, ""),
+                DeviceUsageKind::Configuration(vdev_id) => {
+                    write!(f, "C:{}:{}", vdev_id, self.usage_name)
+                }
+                DeviceUsageKind::Instance => write!(f, "{}", self.usage_name),
+            }
+        }
+    }
+
+    impl std::str::FromStr for DeviceUsage {
+        type Err = ParseDeviceUsageError;
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            if s.is_empty() {
+                return Ok(DeviceUsage {
+                    kind: DeviceUsageKind::Free,
+                    usage_name: s.to_string(),
+                });
+            }
+
+            // Format "C:<vdev_id>:<usage_name>"
+            if let Some((vdev_id, usage_name)) =
+                s.strip_prefix("C:").and_then(|s| s.split_once(':'))
+            {
+                if usage_name.is_empty() {
+                    return Err(ParseDeviceUsageError);
+                }
+                return Ok(DeviceUsage {
+                    kind: DeviceUsageKind::Configuration(vdev_id.to_string()),
+                    usage_name: usage_name.to_string(),
+                });
+            }
+
+            // Format "<usage_name>"
+            Ok(DeviceUsage {
+                kind: DeviceUsageKind::Instance,
+                usage_name: s.to_string(),
+            })
+        }
+    }
+
+    impl DeviceUsage {
+        pub fn create(kind: &DeviceUsageKind, usage_name: &str) -> Result<Self, anyhow::Error> {
+            match kind {
+                DeviceUsageKind::Free => {
+                    if !usage_name.is_empty() {
+                        return Err(anyhow::anyhow!(
+                            "Invalid input parameter, usage name: {} provided for free device usage",
+                            usage_name
+                        ));
+                    };
+                }
+                _ => {
+                    if usage_name.is_empty() {
+                        return Err(anyhow::anyhow!(
+                            "Invalid input parameter, no usage name provided for device usage"
+                        ));
+                    };
+                }
+            };
+
+            Ok(Self {
+                kind: kind.clone(),
+                usage_name: usage_name.into(),
+            })
+        }
+
+        pub fn get_kind(&self) -> DeviceUsageKind {
+            self.kind.clone()
+        }
+
+        pub fn get_usage_name(&self) -> String {
+            self.usage_name.clone()
+        }
+
+        pub fn is_same_usage(&self, usage_name: &str) -> bool {
```

Comment:
should `is_same_usage` depend on the usage kind in addition to the usage name?

maybe this should be renamed to used_by_same_node?

Reply:
no, is_same_usage compare the usage only.  the usage can be any attribute, not limited to node_name.

usage can be any attribute that's referred to the usage of the slot. currently, it can be
node_name: when refers to the slot usage in akrii
slot_id: when refers to the slot usage in annotation

Reply:
oh.  that is not an intuitive overload of behavior to me.  i'll have to take another look.

Reply:
i see the annotation code:

```rust
let slot_usage =
    DeviceUsage::create(&DeviceUsageKind::Instance, &device_usage_id).unwrap();
akri_annotations.insert(
    format!("{}{}", AKRI_SLOT_ANNOTATION_NAME_PREFIX, &device_usage_id),
    slot_usage.to_string(),
);
```

given that the device_usage_id is part of the annotation name, why put it into the DeviceUsage object as well?  why not put the node_name in the DeviceUsage object and be consistent with the slot usage?

Reply:
given the annotation name and value are using the same device_usage_id, is there a time where the annotation name and value won't be alligned?

where this is false: `DeviceUsage::from_str(value).is_same_usage(slot_id)` in crictl_containers.rs:

```rust
if key.starts_with(AKRI_SLOT_ANNOTATION_NAME_PREFIX) {
    let slot_id = key
        .strip_prefix(AKRI_SLOT_ANNOTATION_NAME_PREFIX)
        .unwrap_or_default();
    match DeviceUsage::from_str(value) {
        Ok(slot_usage) => {
            if slot_usage.is_same_usage(slot_id) {
                Some(value.clone())
            } else {
                None
            }
        }
        Err(_) => None,
    }
} else {
    None
}
```

Reply:
nevermind ... seems to match the original code where we were checking annotation name against the value.